R/helpers.R

```diff
@@ -26,6 +26,54 @@ groups_by_filter <- function(covariate, nbins, ties.method="random", seed=NULL){
 	as.factor(ceiling( rfs* nbins))
 }
 
+#' Stratify hypotheses based on increasing value of the covariate 
+#'
+#'  Hypotheses are stratified into nbins different strata of (approximately) equal size based on
+#' increasing value of the covariate
+#' extension of `groups_by_filter` for multidimensional covariates, per-covariate quantiles
+#'
+#' @param covariates Numeric vector of ordinal covariates based on which the stratification will be done.
+#' @param nbins Integer, number of groups/strata into which p-values will be split based on covariate.
+#' @param ties.method Character specifying how ties are treated, see \code{\link{rank}} function.
+#' @param seed Integer, specifies random seed to be used when ties.method=="random".
+#' @return A factor with nbins different levels, each entry corresponds to the stratum the i-th hypothesis
+#'  was assigned to.
+#' @examples
+#' covariates <- matrix(runif(300), ncol = 3)
+#' groups <- groups_by_filter_multivariate(covariates, 10)
+#' @export
+groups_by_filter_multivariate <- function(covariates, nbins, ties.method = "random", seed = NULL) {
+  if (!is.null(seed) && ties.method == "random") {
+    # http://stackoverflow.com/questions/14324096/setting-seed-locally-not-globally-in-r?rq=1
+    tmp <- stats::runif(1)
+    old <- .Random.seed
+    on.exit({
+      .Random.seed <<- old
+    })
+    set.seed(as.integer(seed))
+  }
+  
+  nvar <- ncol(covariates)
+  groups <- lapply(seq_len(nvar), function(i) {
+    covariate_i <- covariates[, i]
```

Comment:
You have to be careful here. I assume you want your method to handle any kind of typical two-dimensional data structure (`matrix`, `data.frame`, and `tibble`). However tibbles behave a bit different: they don't automatically return a vector if you extract a single column:
``` r
tib <- tibble(a = letters[1:5], num = 1:5)
tib[,1]
#> # A tibble: 5 × 1
#>   a    
#>   <chr>
#> 1 a    
#> 2 b    
#> 3 c    
#> 4 d    
#> 5 e
tib[,1,drop=TRUE]
#> [1] "a" "b" "c" "d" "e"
```